src/app/clusters/mode-select-server/mode-select-delegate.h

```diff
@@ -0,0 +1,113 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#ifndef SOME_UNIQUE_NAME_HERE
+#define SOME_UNIQUE_NAME_HERE
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/util/util.h>
+#include <utility>
+#include <vector>
+
+using chip::Protocols::InteractionModel::Status;
+using ModeOptionStructType  = chip::app::Clusters::ModeSelect::Structs::ModeOptionStruct::Type;
+using SemanticTagStructType = chip::app::Clusters::ModeSelect::Structs::SemanticTagStruct::Type;
+
+template <typename T>
+using List = chip::app::DataModel::List<T>;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ModeSelect {
+
+class Delegate
+{
+public:
+    const std::vector<ModeOptionStructType> modeOptions;
+
+    /**
+     * This is a helper function to build a mode option structure. It takes the label/name of the mode,
+     * the value of the mode and a list of semantic tags that apply to this mode.
+     */
+    static ModeOptionStructType BuildModeOptionStruct(const char * label, uint8_t mode,
+                                                      const List<const SemanticTagStructType> & semanticTags)
+    {
+        Structs::ModeOptionStruct::Type option;
+        option.label        = CharSpan::fromCharString(label);
+        option.mode         = mode;
+        option.semanticTags = semanticTags;
+        return option;
+    }
+
+    explicit Delegate(std::vector<ModeOptionStructType> modes) : modeOptions(std::move(modes)) {}
+
+    virtual CHIP_ERROR Init() = 0;
+
+    /**
+     * This function returns true if the mode value given matches one of the supported modes, otherwise it returns false.
+     *
+     * @param mode
+     */
+    bool IsSupportedMode(uint8_t mode);
+
+    /**
+     * If the mode value given is supported, this function will update modeOption to the ModeOptionStructType for the mode with
+     * the matching value and returns success. If the mode value given is not supported, the modeOption remains untouched and
+     * returns an InvalidCommand status.
+     *
+     * @param mode
+     * @param modeOption
+     */
+    Status GetMode(uint8_t mode, ModeOptionStructType & modeOption);
+
+    /**
+     * When a ChangeToMode command is received, if the NewMode value is a supported made, this function is called to decide if we
+     * should go ahead with transitioning to this mode. If this function returns a success status, the change request is accepted
+     * and the CurrentMode is set to the NewMode. Else, the CurrentMode is left untouched and the returned status in added to the
+     * CommandHandler.
+     *
+     * This function is to be overridden by a user implemented function that makes this decision based on the application logic.
+     * @param mode
+     */
+    virtual Status HandleChangeToMode(uint8_t mode);
```

Comment:
Why do we need this?

It seems to me that for the delegate, if we are going to require implementing `HandleChangeToModeWithStatus` (not quite spelled the way you have it) anyway, we might as well _only_ require `HandleChangeToModeWithStatus`.  Then to implement what is currently `HandleChangeToMode` but just in the cluster impl we:

1. Put a `ChangeToModeResponse::Type` on the stack.
2. Call `HandleChangeToModeWithStatus`.
3. If status is UnsupportedMode, respond with INVALID_COMMAND.
4. If status is not Success, respond with FAILURE.
5. Respond with SUCCESS.

Please double-check, but I believe this should correctly implement the ChangeToMode spec semantics.